File: tmp/submission-build/render_pdf_pages.m
#import <Foundation/Foundation.h>
#import <AppKit/AppKit.h>
#import <PDFKit/PDFKit.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        if (argc != 3) {
            fprintf(stderr, "Usage: render_pdf_pages input.pdf output_dir\n");
            return 2;
        }

        NSString *inputPath = [NSString stringWithUTF8String:argv[1]];
        NSString *outputPath = [NSString stringWithUTF8String:argv[2]];
        NSError *directoryError = nil;
        [[NSFileManager defaultManager] createDirectoryAtPath:outputPath
                                  withIntermediateDirectories:YES
                                                   attributes:nil
                                                        error:&directoryError];
        if (directoryError) {
            fprintf(stderr, "Cannot create output directory: %s\n",
                    directoryError.localizedDescription.UTF8String);
            return 3;
        }

        PDFDocument *document = [[PDFDocument alloc] initWithURL:[NSURL fileURLWithPath:inputPath]];
        if (!document) {
            fprintf(stderr, "Cannot open PDF: %s\n", inputPath.UTF8String);
            return 4;
        }

        for (NSInteger index = 0; index < document.pageCount; index++) {
            PDFPage *page = [document pageAtIndex:index];
            NSRect bounds = [page boundsForBox:kPDFDisplayBoxMediaBox];
            NSSize size = NSMakeSize(bounds.size.width * 1.5, bounds.size.height * 1.5);
            NSImage *image = [page thumbnailOfSize:size forBox:kPDFDisplayBoxMediaBox];
            NSData *tiff = image.TIFFRepresentation;
            NSBitmapImageRep *bitmap = [NSBitmapImageRep imageRepWithData:tiff];
            NSData *png = [bitmap representationUsingType:NSBitmapImageFileTypePNG properties:@{}];
            NSString *fileName = [NSString stringWithFormat:@"page-%03ld.png", (long)index + 1];
            [png writeToFile:[outputPath stringByAppendingPathComponent:fileName] atomically:YES];
        }

        printf("%s: %ld pages\n", inputPath.lastPathComponent.UTF8String, (long)document.pageCount);
    }
    return 0;
}

File: tmp/submission-build/render_pdf_pages.swift
import Foundation
import PDFKit
import AppKit

guard CommandLine.arguments.count == 3 else {
    fputs("Usage: render_pdf_pages.swift input.pdf output_dir\n", stderr)
    exit(2)
}

let input = URL(fileURLWithPath: CommandLine.arguments[1])
let outputDir = URL(fileURLWithPath: CommandLine.arguments[2], isDirectory: true)
try FileManager.default.createDirectory(at: outputDir, withIntermediateDirectories: true)

guard let document = PDFDocument(url: input) else {
    fputs("Cannot open PDF: \(input.path)\n", stderr)
    exit(3)
}

for index in 0..<document.pageCount {
    guard let page = document.page(at: index) else { continue }
    let box = page.bounds(for: .mediaBox)
    let scale: CGFloat = 1.5
    let size = NSSize(width: box.width * scale, height: box.height * scale)
    let image = NSImage(size: size)
    image.lockFocus()
    NSColor.white.setFill()
    NSRect(origin: .zero, size: size).fill()
    guard let context = NSGraphicsContext.current?.cgContext else {
        image.unlockFocus()
        continue
    }
    context.saveGState()
    context.scaleBy(x: scale, y: scale)
    page.draw(with: .mediaBox, to: context)
    context.restoreGState()
    image.unlockFocus()

    guard
        let tiff = image.tiffRepresentation,
        let bitmap = NSBitmapImageRep(data: tiff),
        let png = bitmap.representation(using: .png, properties: [:])
    else { continue }

    let name = String(format: "page-%03d.png", index + 1)
    try png.write(to: outputDir.appendingPathComponent(name))
}

print("\(input.lastPathComponent): \(document.pageCount) pages")

